### PR TITLE
Retention email

### PR DESCRIPTION
**Pull request by AaronJackson**
adds two new meta parameters

- `membership_retention_email_defer` defaults to P14D
- `membership_retention_email_subject` defaults to "Hello from Nottingam Hackspace" but membership team might be able to think of something better

can be executed with artisan hms:membership:retention but will result in duplicates if called multiple times on the same day. Scheduled to run daily at 11am.

membership team haven't finalised the email yet, probably best to leave the feature disabled for now.

https://wiki.nottinghack.org.uk/wiki/New_Member_Followup

### CODE REVIEW

**Comment by dpslwk** [comment]
you could add a rate limit, for say 23 hours
https://laravel.com/docs/9.x/queues#rate-limiting

**Comment by AaronJackson** [comment]
hmm dont really understand where to put it. i thought maybe in AppServiceProvider boot but can't get it to fire

```php
        RateLimiter::for('console', function (object $job) {
            print_r($job);
            return Limit::perHour(1);
        });
```

**Comment by dpslwk** [comment]
Yeah the `Ratelimiter::for` bit goes in appserviceprovider boot

then add the middleware section (with the `->doneRelease()` to the job

might need to `php artisan hot:ter` is likley needed to pick up changes

**Comment by dpslwk** [comment]
Your not likely to see that print_r anywhere. `\Log::debug();` and see files can be more useful

**Comment by AaronJackson** [comment]
Cheers, makes sense and seems to work :)

**Comment by AaronJackson** [comment]
Forgot until now that there should probably be a check to make sure they are a current member. I guess it's possible that the interval is set to longer than a month

File: .env.example [hunk 1]
```diff
@@ -117,6 +117,7 @@ MIX_STRIPE_KEY="${STRIPE_KEY}"
 #FEATURE_MW_AUTH_WIKI=false
 #FEATURE_SLACK=true
 #FEATURE_DISCORD=true
+#FEATURE_RETENTION_EMAIL=false
```
Comment (dpslwk) [inline .env.example:120]: this should be the inverse of the default in `config/hms.php`

File: app/Mail/Membership/Retention.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        $subject = $metaRepository->get('membership_retention_email_subject', 'Hello from Nottingham Hackspace');
+        $membershipTeam = $roleRepository->findOneByName('team.membership');
+        $membershipEmail = $membershipTeam->getEmail();
+        $membershipMgEmail = preg_replace('/@/m', '@mg.', $membershipEmail);
+        $displayName = config('branding.space_name') . ' - ' . $membershipTeam->getDisplayName();
+
+        return $this
+            ->replyTo($membershipEmail, $displayName)
+            ->subject($subject)
```
Comment (dpslwk) [inline app/Mail/Membership/Retention.php:52]: All email subjects must follow the same pattern, 

```suggestion
            ->subject(config('branding.space_name') . ': ' . $subject)
```

so the default in meta my want updating to something the does note endup with "Nottingham Hackspace: Hello from Nottingham Hackspace"

File: app/Mail/Membership/Retention.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+     * @param RoleRepository $roleRepository
+     *
+     * @return $this
+     */
+    public function build(MetaRepository $metaRepository, RoleRepository $roleRepository)
+    {
+        $subject = $metaRepository->get('membership_retention_email_subject', 'Hello from Nottingham Hackspace');
+        $membershipTeam = $roleRepository->findOneByName('team.membership');
+        $membershipEmail = $membershipTeam->getEmail();
+        $membershipMgEmail = preg_replace('/@/m', '@mg.', $membershipEmail);
```
Comment (dpslwk) [inline app/Mail/Membership/Retention.php:47]:
```suggestion
```
no used

File: app/Jobs/Membership/RetentionEmailJob.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+            $to = [
+                ['email' => $user->getEmail()],
+            ];
+
+            Mail::to($to)->send(new Retention($user));
+        }
+    }
+
+    public function middleware()
```
Comment (dpslwk) [inline app/Jobs/Membership/RetentionEmailJob.php:63]: preference would be to mode this before the handle method

File: app/Mail/Membership/Retention.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        $displayName = config('branding.space_name') . ' - ' . $membershipTeam->getDisplayName();
+
+        return $this
+            ->replyTo($membershipEmail, $displayName)
```
Comment (dpslwk) [inline app/Mail/Membership/Retention.php:51]: our form name is alread, so not sure this is really needed
Nottinghack Membership Team <membership@nottinghack.org.uk>

File: app/HMS/Repositories/Doctrine/DoctrineProfileRepository.php [hunk 6]
```diff
@@ -139,6 +140,25 @@ public function findOneByDiscordUserSnowflake(string $discordUserSnowflake)
...
+     */
+    public function findByJoinedOn(Carbon $joinDate)
+    {
+        $q = parent::createQueryBuilder('profile');
+
+        $q->where('profile.joinDate >= :join_date_lower')
+          ->andWhere('profile.joinDate < :join_date_upper');
+
+        $q = $q->setParameter('join_date_lower', $joinDate)
+               ->setParameter('join_date_upper', $joinDate->copy()->addDays(1));
```
Comment (dpslwk) [inline app/HMS/Repositories/Doctrine/DoctrineProfileRepository.php:156]: if we add `'DATE' => DoctrineExtensions\Query\Mysql\Date::class,` to `doctrine.custom_datetime_functions` config

we should be able to just do 

```suggestion
        $q->where('DATE(profile.joinDate) = :join_date');

        $q = $q->setParameter('join_date', $joinDate->toDateString());
```

**Comment by AaronJackson** [comment]
cheers matt. didn't know about the custom doctrine functions, much cleaner to do a cast. 
i've rebased this, mainly to make sure stuff is still working after laravel upgrades

File: app/Jobs/Membership/RetentionEmailJob.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+            if (! $user->hasRoleByName(Role::MEMBER_CURRENT)) {
+                continue;
+            }
+
+            $to = [
+                ['email' => $user->getEmail()],
+            ];
+
+            Mail::to($to)->send(new Retention($user));
```
Comment (dpslwk) [inline app/Jobs/Membership/RetentionEmailJob.php:66]:
```suggestion
            Mail::to($user)->send(new Retention($user));
```

**Review by dpslwk — APPROVED** [review]

**Pull request merged**